### PR TITLE
Limit series

### PR DESCRIPTION
**Pull request by shanson7**
This is to prevent OOM issues when users (purposefully or accidentally) ask for too many time-series. From some tests on our side, the size of a time-series response is about 450 bytes * number of series, so accounting for all 250k (the default) would be about 112MB of just raw response buffer, plus the decoded size, plus the capacity overhead (since the responses buffers aren't precisely size) is probably about 250MB (1kB per series). At that point, we just give up. 

The next step is more efficiently decode the peer responses in a streaming manner directly into the in-memory struct (msgp supports it, but we have to worry about the body being closed so I haven't found a clean way yet). At any rate, that would likely be a larger change, so I went with just limiting it first, and making it better second.

### CODE REVIEW

File: api/cluster.go [hunk 1]
```diff
@@ -294,7 +294,7 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
 	result := make(map[string]PeerResponse)
 	for resp := range responses {
 		if resp.err != nil {
-			return nil, err
+			return nil, resp.err
```
Comment (shanson7) [inline api/cluster.go]: This is unrelated boyscouting

**Comment by shanson7** [comment]
@saheemg - This is slightly different async approach but I think the resulting code is quite nice.

File: api/cluster.go [hunk 2]
```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
...
+	result := make(map[string]PeerResponse)
 
-	reqCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	responseChan, errorChan := s.peerQuerySpeculativeChan(ctx, data, name, path)
 
-	originalPeers := make(map[string]struct{}, len(peerGroups))
-	receivedResponses := make(map[int32]struct{}, len(peerGroups))
+	for resp := range responseChan {
+		result[resp.peer.GetName()] = resp
```
Comment (saheemg) [inline api/cluster.go]: peer can only ever send one response?

File: api/graphite.go [hunk 3]
```diff
@@ -675,7 +676,7 @@ func (s *Server) executePlan(ctx context.Context, orgId uint32, plan expr.Plan)
 			for i, e := range exprs {
 				exprs[i] = strings.Trim(e, " '\"")
 			}
-			series, err = s.clusterFindByTag(ctx, orgId, exprs, int64(r.From))
+			series, err = s.clusterFindByTag(ctx, orgId, exprs, int64(r.From), maxSeriesPerReq-len(reqs))
```
Comment (saheemg) [inline api/graphite.go]: where is `reqs` updated?

File: cluster/node.go [hunk 4]
```diff
@@ -201,7 +202,8 @@ func (n HTTPNode) GetName() string {
 func handleResp(rsp *http.Response) ([]byte, error) {
 	defer rsp.Body.Close()
 	if rsp.StatusCode != 200 {
-		ioutil.ReadAll(rsp.Body)
+		// Read in body so that the connection can be reused
+		io.Copy(ioutil.Discard, rsp.Body)
```
Comment (saheemg) [inline cluster/node.go]: why didn't you do this for `rsp.StatusCode == 200` branch?

File: api/cluster.go [hunk 5]
```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
...
+	resultChan := make(chan PeerResponse)
+	errorChan := make(chan error, 1)
 
-		select {
-		case <-ctx.Done():
-			return
-		default:
-			// Not canceled, continue
-		}
+	go func() {
```
Comment (saheemg) [inline api/cluster.go]: I think this  func is like a lambda?  Its kind of long.  Would there be value in refactoring into smaller methods called from the lambada?

(although it looks like longer functions are the norm in the code base)

File: api/cluster.go [hunk 6]
```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
...
+
+		var ticker *time.Ticker
+		var tickChan <-chan time.Time
+		if speculationThreshold != 1 {
+			ticker = time.NewTicker(5 * time.Millisecond)
+			tickChan = ticker.C
+			defer ticker.Stop()
+		}
+
+		for len(receivedResponses) < len(peerGroups) {
```
Comment (saheemg) [inline api/cluster.go]: why `for` here?

File: api/cluster.go (hunk 2 above)
Comment (shanson7) [reply api/cluster.go]: It *should* only send one, as we only sent one request.

File: api/graphite.go (hunk 3 above)
Comment (shanson7) [reply api/graphite.go]: https://github.com/bloomberg/metrictank/pull/36/files/b6b419db4d99c11486ad0e9263cc13570b659504#diff-841baed3f3a566ff48a5a4a427defb30R708

File: cluster/node.go (hunk 4 above)
Comment (shanson7) [reply cluster/node.go]: This line of code reads in and *discards* everything in the response body so that the connection is clean for reuse. in the 200 case, we want that data!

File: api/cluster.go (hunk 5 above)
Comment (shanson7) [reply api/cluster.go]: It is like a lambda, and behaves as a closure (captures scope vars that are referenced). It may be worth refactoring? not sure.

File: api/cluster.go (hunk 6 above)
Comment (shanson7) [reply api/cluster.go]: To loop! `for` is go's `while`

File: cluster/node.go (hunk 4 above)
Comment (saheemg) [reply cluster/node.go]: oh.  I thought maybe there would be a io.Copy call that could help the connection to be reused.

File: api/cluster.go (hunk 2 above)
Comment (saheemg) [reply api/cluster.go]: is worth a check?

Comment (shanson7) [reply api/cluster.go]: Not IMO.

File: api/cluster.go (hunk 6 above)
Comment (saheemg) [reply api/cluster.go]: silly me.  when  I see `for` I expect some sort of range or update  at the declaration point.  for some reason I read this like it was a simple `if`.
